frontend/src/metabase/dashboard/components/PublicLinkPopover/PublicLinkPopover.tsx

```diff
@@ -0,0 +1,124 @@
+import { useState } from "react";
+import { useAsync } from "react-use";
+import { t } from "ttag";
+import {
+  LinkContainer,
+  PublicLinkCopyButton,
+  PublicLinkTextContainer,
+} from "metabase/dashboard/components/PublicLinkPopover/PublicLinkPopover.styled";
+import { useSelector } from "metabase/lib/redux";
+import type { exportFormats } from "metabase/lib/urls";
+import { ExtensionOption } from "metabase/public/components/widgets/SharingPane.styled";
+import { getUserIsAdmin } from "metabase/selectors/user";
+import { Anchor, Box, Group, Popover, Text, Title, Tooltip } from "metabase/ui";
+
+export type ExportFormatType = typeof exportFormats[number] | null;
+
+export type PublicLinkPopoverProps = {
+  target: JSX.Element;
+  isOpen: boolean;
+  onClose: () => void;
+  createPublicLink: () => Promise<void>;
+  deletePublicLink: () => void;
+  uuid: string | null;
+  getPublicLink: ({
+    exportFormat,
+  }: {
+    exportFormat: ExportFormatType;
+  }) => string | null;
+  extensions?: ExportFormatType[];
+};
+
+export const PublicLinkPopover = ({
+  target,
+  createPublicLink,
+  deletePublicLink,
+  uuid,
+  getPublicLink,
+  isOpen,
+  onClose,
+  extensions = [],
+}: PublicLinkPopoverProps) => {
+  const isAdmin = useSelector(getUserIsAdmin);
+
+  const [exportFormat, setExportFormat] = useState<ExportFormatType | null>(
+    null,
+  );
+
+  const url = getPublicLink({ exportFormat });
+
+  const { loading } = useAsync(async () => {
+    if (isOpen && !uuid) {
+      await createPublicLink();
```

Comment:
instead of returning something we could make this function non async and

```ts
return createPublicLink()
```

I haven't tested it, but it should work. Given `async/await` and Promise are isomorphic

Reply:
Ok I have two options for you then - here's what it looks like when we use `return createPublicLink`:

```
  const { loading } = useAsync(() => {
    if (isOpen && !uuid) {
      return createPublicLink();
    }
    
	// we'd need to return a promise since TS won't accept a return of just `undefined`
    return Promise.resolve();
  }, [uuid, isOpen]);
```

If we use `async` here we can get rid of that weird last line, since the type of the function can now be `() => Promise<void|null>`

```
const { loading } = useAsync(async () => {
    if (isOpen && !uuid) {
      return createPublicLink();
    }
  }, [uuid, isOpen]);
```

Let me know what you think. I think the second one is cleaner

Reply:
Yep. The tldr is basically that, without the `async` things aren't automatically wrapped in promises anymore, so option 1's return value is actually `Promise<void> | undefined`  instead of `Promise<void | undefined>` and `useAsync` needs a promise for all possible returns